### PR TITLE
Documentation: DevGuide - transactions

### PR DESCRIPTION
**Pull request by jufajardini**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/4396

Previous PR:
https://github.com/OISF/suricata/pull/5978

Describe changes:
- extending/applayer/transactions: third draft
- fix typo
- add definition of `flow` as found in inliniac's blog
- erase parenthesis text about per-flow state that didn't make much sense
- doc/devguide: add transactions.rst
- doc/devguide/extending/app-layer/index.rst: add transactions.rst

#suricata-verify-pr:
#suricata-verify-repo:
#suricata-verify-branch:
#suricata-update-pr:
#suricata-update-repo:
#suricata-update-branch:
#libhtp-pr:
#libhtp-repo:
#libhtp-branch:

### CODE REVIEW

**Review by catenacyber — COMMENTED** [review]
Transactions progress are also used for certain keywords to know what the minimum state is before we can expect a match.
Before this state, we do not even try to look for the patterns...
cf `DetectAppLayerInspectEngineRegister2` and `DetectAppLayerMpmRegister2`

**Comment by jufajardini** [comment]
Thanks, I'll check those and add that info. :)

File: doc/devguide/extending/app-layer/transactions.rst [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        .
+        .
+    }
+
+_`Progress Tracking`
+====================
+
+As a rule of thumb, transactions will follow a request-response model: if a transaction has had a request and a response, it is complete.
+
+But if a protocol has situations where a request or response won’t expect or generate a message from its counter-part,
```
Comment (jlucovsky) [inline doc/devguide/extending/app-layer/transactions.rst:69]: nit: s/counter-part/counterpart/

File: doc/devguide/extending/app-layer/transactions.rst [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+This is controlled by implementing states. In Suricata, those will be enums that are incremented as the parsing
+progresses. A state will start at 0. The higher its value, the closer the transaction would be to completion.
+
+The engine interacts with transactions state using a set of callbacks the parser registers. State is defined per flow direction (``STREAM_TOSERVER`` / ``STREAM_TOCLIENT``).
+
+In Summary - Transactions and State
+-----------------------------------
+
+| Initial state value: ``0``
```
Comment (jlucovsky) [inline doc/devguide/extending/app-layer/transactions.rst:87]: Suggest a list (bullet, dashed, numbered).

**Comment by jufajardini** [comment]
Followed by: https://github.com/OISF/suricata/pull/6162

**Pull request closed without merging**